Add separate cluster for configured kiali instance

nrfox opened the pull request:
** Describe the change **

This change lets you manually specify a cluster in the kiali config that there is no remote secret for. This cluster is considered "inaccessible" meaning that Kiali won't try to connect to it and the frontend won't consider it for the "multi-cluster mode" e.g. there won't be a cluster column when there's one accessible cluster and one inaccessible cluster.

To test:
1. Create two clusters. You can use: `hack/run-integration-tests.sh --test-suite frontend-multi-primary --setup-only true`.
2. Deploy kiali on both clusters including the "west" cluster. `./hack/istio/multicluster/deploy-kiali.sh --single-kiali false --cluster1-context kind-east --cluster2-context kind-west --kiali-auth-strategy anonymous --manage-kind true --kiali-use-dev-image true`
3. Add the "east" cluster to the "west" kiali config by editing the "west" kiali configmap (`kubectl --context kind-west edit cm kiali -n istio-system`):
    ```
    clustering:
      clusters:
        - name: east
          accessible: false
      kiali_urls:
        - cluster_name: east
          url: http://<kiali_url>/kiali
          instance_name: kiali
          namespace: istio-system
    ```
4. Restart the west kiali: `kubectl --context kind-west delete pod -n istio-system -l app=kiali`
5. Navigate to the west kiali's "mesh" page (you can port-forward to it via `hack/kiali-port-forward.sh -kc kind-west`).
  You should see two kiali instances on the mesh page and a link to the "east" kiali:
![Screenshot from 2023-12-20 13-31-19](https://github.com/kiali/kiali/assets/6226732/e3dc7efe-667e-4700-8ace-a056e69100e0)
  You should _not_ see a cluster column on any of the app/service/istio/workload pages for the "west" kiali.

** Issue reference **

Fixes #6243

nrfox commented:
This kind of works but adds some weird behavior on the frontend that will need to get worked out before this can be merged.

jmazzitelli commented:
I believe you need to set KIALI_USE_DEV_IMAGE before running the test procedure because we want the kiali server built from this PR, not the latest image from quay. So step 2 I think should be:

First, `export KIALI_USE_DEV_IMAGE=true` and then Deploy kiali on both clusters including the "west" cluster. `./hack/istio/multicluster/deploy-kiali.sh --single-kiali false --cluster1-context kind-east --cluster2-context kind-west --kiali-auth-strategy anonymous --manage-kind true`
```

jmazzitelli commented:
step 3 - you need to know to pass `--context kind-west` in order to edit the correct configmap:

`kubectl --context kind-west edit cm kiali -n istio-system`

jmazzitelli commented:
I don't see it  - I'll check my setup to make sure it is as expected:
![image](https://github.com/kiali/kiali/assets/2029470/4c634084-ac5d-4f5a-bb46-dc426d1c678d)

```
kubectl --context kind-west -n istio-system get cm kiali -oyaml
```
shows:
```yaml
    kiali_feature_flags:
...
      clustering:
        autodetect_secrets:
          enabled: true
          label: kiali.io/multiCluster=true
        clusters:
        - name: east
          accessible: false
        kiali_urls:
        - cluster_name: east
          url: http://east-kiali/kiali
          name: kiali
          namespace: istio-system
```

jmazzitelli commented:
That is going to need a release notes blurb - "deprecated: feature flags.clustering has moved up one level, the feature flags.clustering will be removed in a future release."

jmazzitelli commented:
I see this now, but notice I don't see the instance name in the Kiali column. I just see `istio-system/`.

![image](https://github.com/kiali/kiali/assets/2029470/66075a88-ffee-4209-bf91-ef6ddad2b5e4)

The relevant kiali CR section:
```yaml
    clustering:
      autodetect_secrets:
        enabled: true
        label: kiali.io/multiCluster=true
      clusters:
      - name: east
        accessible: false
      kiali_urls:
      - cluster_name: east
        url: http://localhost:20001/kiali
        name: kiali
        namespace: istio-system
```

nrfox commented:
> I see this now, but notice I don't see the instance name in the Kiali column. I just see `istio-system/`.
> 
> ![image](https://private-user-images.githubusercontent.com/2029470/292308794-66075a88-ffee-4209-bf91-ef6ddad2b5e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMxODk5NjAsIm5iZiI6MTcwMzE4OTY2MCwicGF0aCI6Ii8yMDI5NDcwLzI5MjMwODc5NC02NjA3NWE4OC1mZmVlLTQyMDktYmY5MS1lZjZkZGFkMmI1ZTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjFUMjAxNDIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzBhY2QzMDg5YTUwNjg2MzkyNmUwZGU4MTZiZmU5YzQ5M2ZiMzljNzAwNDc0MWU2MjI5NjNhZjIyNWQ5MDM4YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.F6sxOEzrHrTrs4OZUQd-EwFnNXKFJr127DZwdjb2aDI)
> 
> The relevant kiali CR section:
> 
> ```yaml
>     clustering:
>       autodetect_secrets:
>         enabled: true
>         label: kiali.io/multiCluster=true
>       clusters:
>       - name: east
>         accessible: false
>       kiali_urls:
>       - cluster_name: east
>         url: http://localhost:20001/kiali
>         name: kiali
>         namespace: istio-system
> ```

@jmazzitelli I had the config wrong initially in the setup instructions. It should be `instance_name` not `name`.

jmazzitelli commented:
ok. now I see it.
![image](https://github.com/kiali/kiali/assets/2029470/a2e5299e-2525-4b82-ab8f-51e8c3baf1d7)

hhovsepy reviewed (approved):
With the correct location of `clustering:`, running locally:
![Screenshot from 2023-12-21 22-10-34](https://github.com/kiali/kiali/assets/604313/8211b5cf-9ecf-4d3a-bbfd-b328b27ffe81)

nrfox commented:
Added an e2e integration test for this and removed the `Accessible` field for `config.Cluster`. There's now just a `config.Clustering.InaccessibleClusters` field. This is more clear since clusters added manually to the config this way are only for inaccessible clusters and before setting `config.Cluster.Accessible = true` did nothing.

nrfox commented:
Also added a `kiali.Instance` abstraction for the e2e tests that should hopefully make it easier to update the config and restart kiali from within an e2e test.

hhovsepy commented:
I confirm seeing the inaccessible 'east' cluster in mash page with configured link.
Question, if a remote cluster is accessible and not mentioned in `inaccessible_clusters`, it still can be configured by setting a URL, right?

nrfox commented:
> I confirm seeing the inaccessible 'east' cluster in mash page with configured link. Question, if a remote cluster is accessible and not mentioned in `inaccessible_clusters`, it still can be configured by setting a URL, right?

What `URL`? The `clustering.kiali_urls[*].url`? No the way it works is `kiali_url.cluster_name` is just a reference to a cluster. That cluster is defined in `clustering.inaccessible_clusters` or it's an "accessible" cluster that is defined by a cluster secret and kiali will have access to it and those kialis should be automatically discovered so you wouldn't need to add the kiali manually to the config anyways. `kiali_url.cluster_name` should just reference a cluster in `clustering.inaccessible_clusters`. 

I'm not sure I love the name `inaccessible_clusters` but I want the field to convey that these are clusters that kiali doesn't have access to so you need to specify them manually in the config. Hopefully that's clear?

hhovsepy commented:
@nrfox to be more clear, is it possible to override "accessible" cluster's auto discovered Kiali's url by setting `clustering.kiali_urls[*].url` ?

nrfox commented:
> @nrfox to be more clear, is it possible to override "accessible" cluster's auto discovered Kiali's url by setting `clustering.kiali_urls[*].url` ?

Ah ok I haven't tested that out yet but I didn't change that part of the code so it should still work: https://github.com/kiali/kiali/blob/b1dd7b2d392b368059c8f7806edc53273aff6bfa/business/mesh.go#L383-L395

nrfox commented:
Test failure seems to be persistent so I'll have to investigate what's wrong there with `TestAuthPolicyPrincipalsError`.

hhovsepy reviewed (approved):
Regression tested also accessible cluster's discovered Kiali url override via `kiali_urls:` config:
![Screenshot from 2024-01-04 17-55-03](https://github.com/kiali/kiali/assets/604313/9d3c8186-9d72-4b2d-8ca0-ac704854ed1f)
